### PR TITLE
[SPARK-36108][SQL] Refactor first set of 20 query parsing errors to use error classes

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
This PR refactor some exceptions in `QueryParsingErrors` to use error classes.

There are currently ~100 exceptions in this file; so this PR only focuses on the first set of 20.

### Why are the changes needed?
To improve auditing, reduce duplication, and improve quality of error messages thrown from Spark, we should group them in a single JSON file (as discussed in the[ mailing list](http://apache-spark-developers-list.1001551.n3.nabble.com/DISCUSS-Add-error-IDs-td31126.html) and introduced in SPARK-34920).

### Does this PR introduce _any_ user-facing change?
'No'.
Just use new error classes.

### How was this patch tested?
Jenkins test.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @karenfeng FYI

File: core/src/main/resources/error/error-classes.json [hunk 1]
```diff
@@ -63,5 +63,86 @@
...
+  },
+  "UNRECOGNIZED_NOT_MATCHED_ACTION" : {
+    "message" : [ "Unrecognized not matched action: %s" ],
+    "sqlState" : "10007"
+  },
+  "MERGE_STATEMENT_WITHOUT_WHEN_CLAUSE" : {
+    "message" : [ "There must be at least one WHEN clause in a MERGE statement" ],
+    "sqlState" : "10008"
+  },
+  "NON_LAST_MATCHED_CLAUSE_OMIT_CONDITION" : {
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:99]: We can also parametrize these two error messages to simplify auditing.

File: core/src/main/resources/error/error-classes.json [hunk 2]
```diff
@@ -63,5 +63,86 @@
   "WRITING_JOB_ABORTED" : {
     "message" : [ "Writing job aborted" ],
     "sqlState" : "40000"
+  },
+  "INVALID_INSERT_INTO_CONTEXT" : {
+    "message" : [ "Invalid InsertIntoContext" ],
+    "sqlState" : "10001"
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:69]: The SQLSTATEs should reflect those set in the ANSI/ISO standard; see https://github.com/apache/spark/blob/master/core/src/main/resources/error/README.md.

File: core/src/main/resources/error/error-classes.json [hunk 3]
```diff
@@ -63,5 +63,86 @@
...
+  },
+  "INSERT_OVERWRITE_DIRECTORY_UNSUPPORTED" : {
+    "message" : [ "INSERT OVERWRITE DIRECTORY is not supported" ],
+    "sqlState" : "10002"
+  },
+  "COLUMNS_ALIASES_NOT_ALLOWED_IN_OPERATION" : {
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:75]: Grammar nit: `Columns aliases` -> `Column aliases`.

File: core/src/main/resources/error/error-classes.json [hunk 4]
```diff
@@ -63,5 +63,86 @@
...
+    "message" : [ "Columns aliases are not allowed in %s." ],
+    "sqlState" : "10003"
+  },
+  "EMPTY_SOURCE_FOR_MERGE" : {
+    "message" : [ "Empty source for merge: you should specify a source table/subquery in merge." ],
+    "sqlState" : "10004"
+  },
+  "UNRECOGNIZED_MATCHED_ACTION" : {
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:83]: Can we simplify this by parametrizing? Then we only need one for `MATCHED`/`NOT_MATCHED`.

File: core/src/main/resources/error/error-classes.json [hunk 5]
```diff
@@ -63,5 +63,86 @@
...
+  },
+  "TRANSFORM_WITH_SERDE_UNSUPPORTED" : {
+    "message" : [ "TRANSFORM with serde is only supported in hive mode" ],
+    "sqlState" : "10015"
+  },
+  "LATERAL_WITH_PIVOT_IN_FROM_CLAUSE_NOT_ALLOWED" : {
+    "message" : [ "LATERAL cannot be used together with PIVOT in FROM clause" ],
+    "sqlState" : "10016"
+  },
+  "LATERAL_JOIN_WITH_NATURAL_JOIN_UNSUPPORTED" : {
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:131]: I think we can merge these as well.

File: core/src/main/resources/error/error-classes.json [hunk 6]
```diff
@@ -63,5 +63,86 @@
...
+    "message" : [ "Unsupported LATERAL join type %s" ],
+    "sqlState" : "10019"
+  },
+  "INVALID_LATERAL_JOIN_RELATION" : {
+    "message" : [ "LATERAL can only be used with subquery" ],
+    "sqlState" : "10020"
   }
-}
\ No newline at end of file
+}
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:147]: This newline may be causing the test failures; we guarantee that the spacing is correct with a round trip read/write in the unit tests.

**Review by karenfeng — COMMENTED** [review]
Thanks for working on this! I left some suggestions to simplify the error classes. We should also make sure the SQLSTATEs match the ISO standard; if you're not sure what SQLSTATE fits, you can also leave it empty for now.

File: core/src/main/resources/error/error-classes.json [hunk 7]
```diff
@@ -63,5 +63,86 @@
   "WRITING_JOB_ABORTED" : {
     "message" : [ "Writing job aborted" ],
     "sqlState" : "40000"
+  },
+  "INVALID_INSERT_INTO_CONTEXT" : {
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:67]: These need to be in alphabetical order to pass the style tests.

**Comment by beliefer** [comment]
> Thanks for working on this! I left some suggestions to simplify the error classes. We should also make sure the SQLSTATEs match the ISO standard; if you're not sure what SQLSTATE fits, you can also leave it empty for now.

Can you tell me how to know the SQLSTATEs?

File: core/src/main/resources/error/error-classes.json (hunk 2 above)
Comment (beliefer) [reply core/src/main/resources/error/error-classes.json]: I'm not sure.

File: core/src/main/resources/error/error-classes.json (hunk 4 above)
Comment (beliefer) [reply core/src/main/resources/error/error-classes.json]: OK

File: core/src/main/resources/error/error-classes.json (hunk 5 above)
Comment (beliefer) [reply core/src/main/resources/error/error-classes.json]: Yeah.

File: core/src/main/resources/error/error-classes.json (hunk 7 above)
Comment (beliefer) [reply core/src/main/resources/error/error-classes.json]: I got it.

**Comment by karenfeng** [comment]
Hi @beliefer! The SQLSTATEs are in the ISO standard, but it's behind a paywall. You can look at free resources like the [Wikipedia article](https://en.wikipedia.org/wiki/SQLSTATE), or the [Oracle](https://docs.oracle.com/cd/E18283_01/appdev.112/e10827/appd.htm) or [DB2](https://www.ibm.com/docs/en/db2-for-zos/11?topic=codes-sqlstate-values-common-error) manuals (although they often define their own classes and subclasses). I'll push a PR soon with a summarized version of the ISO standards.

File: core/src/main/resources/error/error-classes.json [hunk 8]
```diff
@@ -3,6 +3,18 @@
     "message" : [ "Field name %s is ambiguous and has %s matching fields in the struct." ],
     "sqlState" : "42000"
   },
+  "COLUMN_ALIASES_NOT_ALLOWED_IN_OPERATION" : {
+    "message" : [ "Columns aliases are not allowed in %s." ],
+    "sqlState" : "10000"
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:8]: See https://github.com/apache/spark/pull/33560 for the standard SQLSTATEs. I think this would be 42000; we'll do another pass before the 3.3 release to clean up any incorrect ones.

**Comment by beliefer** [comment]
ping @karenfeng @HyukjinKwon Any other suggestion?

**Review by karenfeng — APPROVED** [review]
Thanks for doing this! LGTM. @HyukjinKwon, can you also take a look?

**Comment by beliefer** [comment]
ping @HyukjinKwon too.

File: core/src/main/resources/error/error-classes.json [hunk 9]
```diff
@@ -3,6 +3,10 @@
     "message" : [ "Field name %s is ambiguous and has %s matching fields in the struct." ],
     "sqlState" : "42000"
   },
+  "COLUMN_ALIASES_NOT_ALLOWED_IN_OPERATION" : {
+    "message" : [ "Columns aliases are not allowed in %s." ],
```
Comment (cloud-fan) [inline core/src/main/resources/error/error-classes.json:7]: `columns aliases` -> `column aliases`?

File: core/src/main/resources/error/error-classes.json [hunk 10]
```diff
@@ -29,10 +41,30 @@
...
+  "INSERTED_VALUE_NUMBER_NOT_MATCH_FIELD_NUMBER" : {
+    "message" : [ "The number of inserted values cannot match the fields." ],
+    "sqlState" : "42000"
+  },
   "INVALID_FIELD_NAME" : {
     "message" : [ "Field name %s is invalid: %s is not a struct." ],
     "sqlState" : "42000"
   },
+  "INVALID_INSERT_INTO_CONTEXT" : {
+    "message" : [ "Invalid InsertIntoContext" ],
```
Comment (cloud-fan) [inline core/src/main/resources/error/error-classes.json:53]: This error message looks a big vague. What exactly the error is?

Comment (beliefer) [reply core/src/main/resources/error/error-classes.json:53]: The message "Invalid InsertIntoContext" is thrown when withInsertInto does not match.
https://github.com/apache/spark/blob/de62b5ae325741e6617e9400f78c6d0fc8cea5de/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala#L267

Comment (cloud-fan) [reply core/src/main/resources/error/error-classes.json:53]: This is a user-facing error message, do we expect end-users to understand this message?

File: core/src/main/resources/error/error-classes.json [hunk 11]
```diff
@@ -57,9 +93,25 @@
     "message" : [ "The second argument of '%s' function needs to be an integer." ],
     "sqlState" : "22023"
   },
+  "TRANSFORM_NOT_SUPPORT_QUANTIFIER" : {
+    "message" : [ "TRANSFORM does not support DISTINCT/ALL in inputs" ],
+    "sqlState" : "42000"
+  },
+  "TRANSFORM_WITH_SERDE_UNSUPPORTED" : {
+    "message" : [ "TRANSFORM with serde is only supported in hive mode" ],
+    "sqlState" : "42000"
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:102]: This should probably be 0A000

File: core/src/main/resources/error/error-classes.json [hunk 12]
```diff
@@ -57,9 +93,25 @@
...
     "message" : [ "Unable to acquire %s bytes of memory, got %s" ]
   },
+  "UNRECOGNIZED_ACTION" : {
+    "message" : [ "Unrecognized %s action: %s" ],
+    "sqlState" : "42000"
+  },
+  "UNSUPPORTED_LATERAL_JOIN_TYPE" : {
+    "message" : [ "Unsupported LATERAL join type %s" ],
+    "sqlState" : "42000"
+  },
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:114]: This should probably be 0A000

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 13]
```diff
@@ -292,8 +292,6 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with SQLConfHelper with Logg
       case hiveDir: InsertOverwriteHiveDirContext =>
         val (isLocal, storage, provider) = visitInsertOverwriteHiveDir(hiveDir)
         InsertIntoDir(isLocal, storage, provider, query, overwrite = true)
-      case _ =>
```
Comment (karenfeng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:295]: Why was this removed?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryParsingErrors.scala [hunk 14]
```diff
@@ -29,91 +29,84 @@ import org.apache.spark.sql.catalyst.trees.Origin
...
-    new ParseException("Empty source for merge: you should specify a source" +
-      " table/subquery in merge.", ctx.source)
+    new ParseException("EMPTY_SOURCE_FOR_MERGE", Array.empty, ctx.source)
   }
 
   def unrecognizedMatchedActionError(ctx: MatchedClauseContext): Throwable = {
-    new ParseException(s"Unrecognized matched action: ${ctx.matchedAction().getText}",
-      ctx.matchedAction())
+    new ParseException("UNRECOGNIZED_ACTION",
+      Array("matched", ctx.matchedAction().getText), ctx.matchedAction())
```
Comment (karenfeng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryParsingErrors.scala:46]: Can we make `matched` all caps? This may make it clearer.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryParsingErrors.scala [hunk 15]
```diff
@@ -29,91 +29,84 @@ import org.apache.spark.sql.catalyst.trees.Origin
...
-    new ParseException("The number of inserted values cannot match the fields.",
+    new ParseException("INSERTED_VALUE_NUMBER_NOT_MATCH_FIELD_NUMBER", Array.empty,
       ctx.notMatchedAction())
   }
 
   def unrecognizedNotMatchedActionError(ctx: NotMatchedClauseContext): Throwable = {
-    new ParseException(s"Unrecognized not matched action: ${ctx.notMatchedAction().getText}",
-      ctx.notMatchedAction())
+    new ParseException("UNRECOGNIZED_ACTION",
+      Array("not matched", ctx.notMatchedAction().getText), ctx.notMatchedAction())
```
Comment (karenfeng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryParsingErrors.scala:56]: Can we make `not matched` all caps? This may make it clearer.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryParsingErrors.scala [hunk 16]
```diff
@@ -29,91 +29,84 @@ import org.apache.spark.sql.catalyst.trees.Origin
...
 
   def emptyPartitionKeyError(key: String, ctx: PartitionSpecContext): Throwable = {
-    new ParseException(s"Found an empty partition key '$key'.", ctx)
+    new ParseException("EMPTY_PARTITION_KEY", Array(key), ctx)
   }
 
   def combinationQueryResultClausesUnsupportedError(ctx: QueryOrganizationContext): Throwable = {
-    new ParseException(
-      "Combination of ORDER BY/SORT BY/DISTRIBUTE BY/CLUSTER BY is not supported", ctx)
+    new ParseException("OPERATION_UNSUPPORTED",
```
Comment (karenfeng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryParsingErrors.scala:76]: We should probably have a separate error class for unsupported combinations.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryParsingErrors.scala [hunk 17]
```diff
@@ -29,91 +29,84 @@ import org.apache.spark.sql.catalyst.trees.Origin
...
   }
 
   def lateralWithPivotInFromClauseNotAllowedError(ctx: FromClauseContext): Throwable = {
-    new ParseException("LATERAL cannot be used together with PIVOT in FROM clause", ctx)
+    new ParseException("LATERAL_WITH_PIVOT_IN_FROM_CLAUSE_NOT_ALLOWED", Array.empty, ctx)
   }
 
   def lateralJoinWithNaturalJoinUnsupportedError(ctx: ParserRuleContext): Throwable = {
-    new ParseException("LATERAL join with NATURAL join is not supported", ctx)
+    new ParseException("OPERATION_UNSUPPORTED", Array("LATERAL join with NATURAL join"), ctx)
```
Comment (karenfeng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryParsingErrors.scala:97]: I think we can also make a separate error class here as well, maybe about how different join types are incompatible.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala [hunk 18]
```diff
@@ -42,6 +42,12 @@ class PlanParserSuite extends AnalysisTest {
   private def intercept(sqlCommand: String, messages: String*): Unit =
     interceptParseException(parsePlan)(sqlCommand, messages: _*)
 
+  private def interceptWithErrorClass(sqlCommand: String, messages: String*)(
```
Comment (karenfeng) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala:45]: Nice test!

File: core/src/main/resources/error/error-classes.json [hunk 19]
```diff
@@ -29,10 +41,26 @@
     "message" : [ "Invalid pivot column '%s'. Pivot columns must be comparable." ],
     "sqlState" : "42000"
   },
+  "INSERTED_VALUE_NUMBER_NOT_MATCH_FIELD_NUMBER" : {
+    "message" : [ "The number of inserted values cannot match the fields." ],
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:45]: Nit: `cannot` -> `does not`

File: core/src/main/resources/error/error-classes.json [hunk 20]
```diff
@@ -29,10 +41,26 @@
     "message" : [ "Invalid pivot column '%s'. Pivot columns must be comparable." ],
     "sqlState" : "42000"
   },
+  "INSERTED_VALUE_NUMBER_NOT_MATCH_FIELD_NUMBER" : {
```
Comment (karenfeng) [inline core/src/main/resources/error/error-classes.json:44]: The grammar is a bit strange. Maybe `INSERTED_VALUE_AND_FIELD_NUMBER_MISMATCH`?

**Comment by beliefer** [comment]
@cloud-fan Could you help me to remove `Stale` label?
